modules/afuser/afuser.c

```diff
@@ -154,14 +145,25 @@ afuser_dd_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_options)
           strncpy(p, ut->ut_line, sizeof(ut->ut_line));
           p[sizeof(ut->ut_line)] = 0;
 
+          /* ut_user/ut_name may not be null-terminated. Copy it and add null-terminator manually. */
+#ifdef SYSLOG_NG_HAVE_MODERN_UTMP
+          gchar user[sizeof(ut->ut_user) + 1]; /* user name, null terminator */
+          strncpy(user, ut->ut_user, sizeof(ut->ut_user));
+          user[sizeof(ut->ut_user)] = 0;
+#else
+          gchar user[sizeof(ut->ut_name) + 1]; /* user name, null terminator */
+          strncpy(user, ut->ut_name, sizeof(ut->ut_name));
+          user[sizeof(ut->ut_name)] = 0;
```

Comment:
Instead of using `evt_tag_str` you could simply use `evt_tag_printf`, which work just like a printf, where you can specify the length of the string (`evt_tag_printf("user", "%.*s", string, string_length)`).

Alternative idea is to create a length restricted version of `evt_tag_str` something like `evt_tag_str_len`, as I see  a few possible candidates for that:

```
lib/logmsg/logmsg.h:522:  return evt_tag_printf(name, "%.*s", (gint) value_len, value);
lib/logproto/logproto-buffered-server.c:140:                            evt_tag_printf("buffer", "%.*s", (gint) state->pending_buffer_end, self->buffer));
lib/logproto/logproto-buffered-server.c:660:                evt_tag_printf("buffer", "%.*s", (gint) buffer_len, buffer));
lib/logproto/logproto-framed-server.c:163:                    evt_tag_printf("header", "%.*s", (gint) (i - self->buffer_pos), &self->buffer[self->buffer_pos]));
lib/logproto/logproto-proxied-text-server.c:97:              evt_tag_printf("parameters", "%.*s", (gint) msg_len, msg));
lib/logproto/logproto-proxied-text-server.c:278:  msg_debug("PROXY protocol header received", evt_tag_printf("line", "%.*s",
lib/logreader.c:454:            evt_tag_printf("input", "%.*s", length, line),
lib/rewrite/rewrite-subst.c:71:                evt_tag_printf("input", "%.*s", (gint) length, value),
lib/rewrite/rewrite-subst.c:83:                evt_tag_printf("input", "%.*s", (gint) length, value),
lib/logmatcher.c:243:                      evt_tag_printf("value", "%.*s", (gint) value_len, value));
lib/logmatcher.c:252:                  evt_tag_printf("value", "%.*s", (gint) value_len, value));
modules/afsmtp/afsmtp.c:361:                 evt_tag_printf("message", "%.*s", buflen, buf));
modules/afsmtp/afsmtp.c:366:                evt_tag_printf("message", "%.*s", buflen, buf));
modules/afsmtp/afsmtp.c:371:                evt_tag_printf("data", "%.*s", buflen, buf));
modules/confgen/confgen-plugin.c:122:            evt_tag_printf("block", "%.*s", (gint) result->len, result->str));
modules/kafka/kafka-dest-driver.c:298:                evt_tag_printf("message", "%.*s", (int) MIN(len, 128), (char *) payload),
modules/kafka/kafka-dest-driver.c:308:                evt_tag_printf("message", "%.*s", (int) MIN(len, 128), (char *) payload),
modules/mqtt/destination/mqtt-destination.c:115:  msg_error("MQTT TLS error", evt_tag_printf("line", "%.*s", (gint) len, str),
modules/mqtt/source/mqtt-source.c:101:  msg_error("MQTT TLS error", evt_tag_printf("line", "%.*s", (gint) len, str),
modules/pseudofile/pseudofile.c:97:  return evt_tag_printf("message", "%.*s%s",
```

Reply:
Ah yea, interesting, I somehow missed `evt_tag_printf()`, but that solves the problem without requiring the copy. How about a wrapper like this:
```
static EVTTAG *
_evt_tag_utmp_username(UtmpEntry *ut)
{
  /* ut_user/ut_name may not be null-terminated */
#ifdef SYSLOG_NG_HAVE_MODERN_UTMP
  return evt_tag_printf("user", "%.*s", (int)sizeof(ut->ut_user), ut->ut_user);
#else
  return evt_tag_printf("user", "%.*s", (int)sizeof(ut->ut_name), ut->ut_name);
#endif
}
```

Reply:
Regarding `evt_tag_str_len`: Looks like there already is `evt_tag_mem` which basically does exactly that (luckily it prints the memory as ASCII string, not hex dump or such).

Reply:
Thanks for the update, it is fine to do the whole project replace. You can just use `refactor:` or `all: ` (see 5ce893d36c17ad6a78d159c1ff710a971619940b) , sure this is not the *module* but the regex really just checks for a structure `something: something more`.